src/ess.jl

```diff
@@ -0,0 +1,353 @@
+# methods
+abstract type AbstractESSMethod end
+
+"""
+    ESSMethod <: AbstractESSMethod
+
+The `ESSMethod` uses a standard algorithm for estimating the
+effective sample size of MCMC chains.
+
+It is is based on the discussion by
+[Vehtari et al. (2019)](https://arxiv.org/pdf/1903.08008.pdf) and uses the
+plug-in estimator of the autocorrelation function discussed by
+[Geyer (1992)](https://projecteuclid.org/euclid.ss/1177011137).
+"""
+struct ESSMethod <: AbstractESSMethod end
+
+"""
+    FFTESSMethod <: AbstractESSMethod
+
+The `FFTESSMethod` uses a standard algorithm for estimating
+the effective sample size of MCMC chains.
+
+It is is based on the discussion by
+[Vehtari et al. (2019)](https://arxiv.org/pdf/1903.08008.pdf) and uses the
+plug-in estimator of the autocorrelation function discussed by
+[Geyer (1992)](https://projecteuclid.org/euclid.ss/1177011137). In contrast to
+`ESSMethod`, it uses fast Fourier transforms (FFTs) for
+estimating the autocorrelation function.
+"""
+struct FFTESSMethod <: AbstractESSMethod end
+
+"""
+    BDAESSMethod <: AbstractESSMethod
+
+The `BDAESSMethod` uses a standard algorithm for estimating the effective sample size of
+MCMC chains.
+
+It is is based on the discussion by
+[Vehtari et al. (2019)](https://arxiv.org/pdf/1903.08008.pdf) and uses the
+estimator of the autocorrelation function discussed in 
+[Bayesian Data Analysis (2013)](https://www.taylorfrancis.com/books/9780429113079).
+"""
+struct BDAESSMethod <: AbstractESSMethod end
+
+# caches
+mutable struct ESSCache{T}
+    samples::Matrix{T}
+    var::Vector{T}
+end
+
+mutable struct FFTESSCache{T,P,I}
+	A::T
```

Comment:
Was my mistake (shouldn't have used tab):
```suggestion
    A::T
```